app/models/user.rb

```diff
@@ -1433,12 +1413,33 @@ def roles
     # array, so that we can append roles which are not yet migrated to the new system. This can be
     # removed once all roles are migrated to the new system.
 
-    if delegate_status.present?
-      roles << delegate_role
-    end
-
     roles.concat(team_roles)
 
     roles
   end
+
+  private def highest_delegate_role
+    highest_status_rank = nil
+    highest_role = nil
+    delegate_roles.each do |role|
+      status_rank = UserRole.status_sort_rank(role)
+      if highest_status_rank.nil? || status_rank < highest_status_rank
+        highest_role = role
+        highest_status_rank = status_rank
+      end
+    end
+    highest_role
+  end
+
+  def delegate_status
+    highest_delegate_role&.metadata&.status
+  end
+
+  def region_id
+    highest_delegate_role&.group&.id
```

Comment:
I feel like `role&.group_id` should be enough, is there an actual need to load the full `group` row into memory?